### ISSUE TITLE
Feature: support ternary operator in structural directive

### ISSUE THREAD

**Issue by MikeDabrowski** [ngneat/transloco-keys-manager#41]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[ ] Documentation issue or request
[ ] Support request
[ ] Other... Please describe:
</code></pre>

## Current behavior

I have this in HTML
```
<app-card *transloco="let t; read: 'CREATE_COMPANY.ADD_LOGO'">
      <!-- t(CROP_IMAGE) -->
      <h1>{{ t(imgRead ? 'CROP_IMAGE' : 'UPLOAD') }}</h1>
```

And in en.json
```
"CREATE_COMPANY": {
  "ADD_LOGO": {
    "CROP_IMAGE": "Crop Image",
  }
}
```
When I run key-manager find I get:
![image](https://user-images.githubusercontent.com/17836911/83741181-2c5f5700-a658-11ea-8ff8-1b7cfc635ea2.png)

Note that for 'de' there are two missing keys, and only the first one is correct - in fact it does not exist in de.json but it is present in en.json.
Same when I do not use ternary there. Key manager duplicates in a way that key. If I add CROP_IMAGE to en.json at top level finder returns all good, but the actual value that is displayed comes from nested CROP_IMAGE not top level one.

## Expected behavior

Key manager finder to properly inherit 'read' and not duplicate keys

## Minimal reproduction of the problem with instructions

Should be easily repro on any code, sadly I don't have time to prep stackblitz.

## Environment

<pre><code>
Angular version: 9.1+,
Keymgr: 2.0.7

For Tooling issues:
- Node version: 12.16.3
- Platform:  Win10

**Comment by shaharkazaz** [comment]
@MikeDabrowski why do you have `<!-- t(CROP_IMAGE) -->`?

**Comment by MikeDabrowski** [comment]
@shaharkazaz I want to extract the keys below that line, isnt that the way to go ?
Without that comment key manager lists that keys in extra column

**Comment by shaharkazaz** [comment]
@MikeDabrowski the comment markings are used for [dynamic keys](https://github.com/ngneat/transloco-keys-manager#dynamic-keys), for regular extraction you shouldn't add them.
Can you remove it and LMK what's the current issue?

**Comment by MikeDabrowski** [comment]
Ok, so I removed the comment, removed  CREATE_COMPANY.ADD_LOGO.CROP_IMAGE and CREATE_COMPANY.ADD_LOGO.UPLOAD from en.json and ran `find`. Here is the result:

```
┌───────────┬──────────────┬──────────────────────────────────────┐
│ File Name │ Missing Keys │ Extra Keys                           │
├───────────┼──────────────┼──────────────────────────────────────┤
│ de        │ --           │ 'CREATE_COMPANY.ADD_LOGO.CROP_IMAGE' │
│           │              │ 'CREATE_COMPANY.ADD_LOGO.UPLOAD'     │
├───────────┼──────────────┼──────────────────────────────────────┤
│ fr        │ --           │ 'CREATE_COMPANY.ADD_LOGO.CROP_IMAGE' │
│           │              │ 'CREATE_COMPANY.ADD_LOGO.UPLOAD'     │
└───────────┴──────────────┴──────────────────────────────────────┘
```
Which tells me that manager could not extract CROP_IMAGE and UPLOAD. Just by adding the comment and reruning `find` i get:
```
┌───────────┬──────────────────────────────────────┬────────────┐
│ File Name │ Missing Keys                         │ Extra Keys │
├───────────┼──────────────────────────────────────┼────────────┤
│ de        │ 'CROP_IMAGE'                         │ --         │
│           │ 'UPLOAD'                             │            │
├───────────┼──────────────────────────────────────┼────────────┤
│ en        │ 'CREATE_COMPANY.ADD_LOGO.CROP_IMAGE' │ --         │
│           │ 'CREATE_COMPANY.ADD_LOGO.UPLOAD'     │            │
│           │ 'CROP_IMAGE'                         │            │
│           │ 'UPLOAD'                             │            │
├───────────┼──────────────────────────────────────┼────────────┤
│ fr        │ 'CROP_IMAGE'                         │ --         │
│           │ 'UPLOAD'                             │            │
└───────────┴──────────────────────────────────────┴────────────┘
```
For me this looks like duplication that may be caused by not prepending the key with read prefix at some point.

**Comment by shaharkazaz** [comment]
@MikeDabrowski I'll try to reproduce but I might need your help with that, I'll get back to you 😃

**Comment by MikeDabrowski** [comment]
Thanks, next week I can create some stackblitz if still needed. I don't have time to do that now, unfortunately.

**Comment by MikeDabrowski** [comment]
Sorry, it took me longer than expected but here is the repo: https://github.com/ngneat/transloco-keys-manager

I expect that putting <!--> comment inside 'read' will result (after extraction) in a nested extracted key. So: I have `*transloco="t(cond ? "A" : "B")"` and inside that tag I put `<!-- t(A, B) -->`. After extract I only have keys A and B flat on top level of json.

When I manually add proper, nested key NEST.A to json it is then properly rendered BUT finder now complains it is extra key.

Pls let me know if this repro is sufficient or maybe am I making a mistake in assuming how it works.

**Comment by shaharkazaz** [comment]
@MikeDabrowski I think you added a wrong link to the repo 🙂 
But just from reading the explanation, I think you are misunderstanding the purpose of the comments, I'll wait for the correct link and see 👍

**Comment by MikeDabrowski** [comment]
@shaharkazaz sorry, copypasta error :P

**Comment by trandromeda** [comment]
I think I'm experiencing something similar to this issue. 

I have this in my en.json:

```
  "landing": {
    "name": "Name",
    "name__error": "Please enter a name",
    "email": "Email",
}
```

and in my template I have used Transloco as such:
```
<ng-container *transloco="let t; read: 'landing'">
  <label for="name">{{ t('name') }}</label>
 ... (and the rest of the keys)
</ng-container>
```

When running the extract command, instead of detecting the nested keys already in my json it creates duplicate keys e.g:
```
{
  "landing.name": "Missing value for 'landing.name'",
  "landing.name__error": "Missing value for 'landing.name__error'",
  "landing.email": "Missing value for 'landing.email'",
  "landing": {
    "name": "Name",
    "name__error": "Please enter a name",
    "email": "Email",
  }
}
```

Perhaps this is intended but it would be nice to be able to nest keys in our json for sure!

**Comment by MikeDabrowski** [comment]
For me, such template will be correctly parsed and put in JSON. I use it
like that all the time. Only combination of dynamic keys and comments has
issues.

**Comment by trandromeda** [comment]
> I think I'm experiencing something similar to this issue.
> 
> I have this in my en.json:
> 
> ```
>   "landing": {
>     "name": "Name",
>     "name__error": "Please enter a name",
>     "email": "Email",
> }
> ```
> 
> and in my template I have used Transloco as such:
> 
> ```
> <ng-container *transloco="let t; read: 'landing'">
>   <label for="name">{{ t('name') }}</label>
>  ... (and the rest of the keys)
> </ng-container>
> ```
> 
> When running the extract command, instead of detecting the nested keys already in my json it creates duplicate keys e.g:
> 
> ```
> {
>   "landing.name": "Missing value for 'landing.name'",
>   "landing.name__error": "Missing value for 'landing.name__error'",
>   "landing.email": "Missing value for 'landing.email'",
>   "landing": {
>     "name": "Name",
>     "name__error": "Please enter a name",
>     "email": "Email",
>   }
> }
> ```
> 
> Perhaps this is intended but it would be nice to be able to nest keys in our json and to also use this awesome tool for sure!

After a closer look I realized I needed to add the `unflat` option when running extract. GL with the original issue being resolved!

**Comment by shaharkazaz** [comment]
@MikeDabrowski 
I think you misunderstood the comments usage.
From your reproduction:
```
<div *transloco="let t; read: 'NEST'">
  <!-- t(NEST_A, NEST_B) -->
  <p>This is the problem: {{t( condition ? 'NEST_A': 'NEST_B')}}</p>
</div>
```
Why was the comment added there? there are no dynamic values here, the keys manager can handle a ternary.
Comments are always added to the top-level and are not influenced by the `read` property.
A good usage for the comments would be:
```
<div *transloco="let t; read: 'NEST'">
  <!-- t(NEST.NEST_A, NEST.NEST_B) -->
  <p>This is the problem: {{t( condition ? variable1: variable2 )}}</p>
</div>
```
`variable1` is `'NEST_A'`, `variable2` is `'NEST_B'` and since we want them under the `'NEST'` we prefix the comment values with it.
I hope it's clearer now 😄

**Comment by MikeDabrowski** [comment]
The comment was added because key manager failed to parse ` <p>This is the problem: {{t( condition ? 'NEST_A': 'NEST_B')}}</p>` and did not put those two into json.  [case2](https://github.com/MikeDabrowski/transloco-km-bug-repro/commit/5472befa19ceb8dc01b375a02661c59a28f70744)
And after adding the comment it failed to nest those two values. [case3](https://github.com/MikeDabrowski/transloco-km-bug-repro/commit/7b6d8b13915679b5c080007a83929a85bcbff98a)
Even when i moved strings to variables it failed to extract them. [case1](https://github.com/MikeDabrowski/transloco-km-bug-repro/commit/d85b9ddc73147fda20d3364fb4a25d47da5ee49a) So I still don't get how it can handle ternary.

Also:

> Comments are always added to the top-level and are not influenced by the read property.

Then this is clearly opposite to what manual (readme) says:

> When using comments in the templates they will also inherit the read input value (if exists), and will be prefixed with it

So what am I still missing here?

**Comment by shaharkazaz** [comment]
@MikeDabrowski Sorry I made a mistake, let me re-clarify: (never write a comment while being distracted haha)
1. Comments should inherit the read value and prefixed by it, I ran the extraction and it works fine IDK why you were using v1 in the repo you provided upgrading made it work correctly.
2. The ternary is currently supported in the pipes and directives (and not the structural) it's shown in the examples but I agree it needs to be written.
3. Case 1 shouldn't have worked since when using variables **you need the comment** to tell the manager what are the possible keys.

Would you like to create a PR and support ternary in the structural directives as well?

**Comment by MikeDabrowski** [comment]
Darn... Those versions are seriously messed up. Sorry about that. I bumped it to 2.2.1 now and rechecked. I can confirm, 2.2.1 nests comments as in readme. I need to keep an eye out and bump versions more often :)

Case 1: I thought so but I guess that was that mistake of yours so we can scratch that.

So what is left is writing the support for ternary in structural, right? I'll be happy to add PR for this but can't promise when. I will try to find time.

**Comment by shaharkazaz** [comment]
@MikeDabrowski I think it's a feature request 😄 
If you think you can make a Pr in the next 2 weeks or so cool! if not I'll add the feature 👍

**Comment by MikeDabrowski** [comment]
2 weeks is doable :D

**Comment by shaharkazaz** [comment]
@MikeDabrowski Great! waiting for that PR 😄

**Comment by MikeDabrowski** [comment]
@shaharkazaz 
One more detail.
I updated km to 2.2.1 and put this snippet in my commercial code:

```
<app-card *transloco="let t; read: 'CREATE_COMPANY.ADD_LOGO'">
      <!-- t(TEST) -->
      <p>{{ t('TEST')}} </p>
      <!-- t(CROP_IMAGE, UPLOAD) -->
      <h1>{{ t(imgRead ? 'CROP_IMAGE' : 'UPLOAD') }}</h1>
```

and key manager find outputs: 

```
│ en        │ 'CREATE_COMPANY.ADD_LOGO.TEST' │ --                                                  │
│           │ 'TEST'                         │                                                     │
│           │ 'CROP_IMAGE'                   │                                                     │
│           │ 'UPLOAD'                       │                                                     │
├───────────┼────────────────────────────────┼─────────────────────────────────────────────────────┤
```

Shouldn't it output only one 'TEST' ? I mean it is not used in ternary, but its both extracted individually and with read input combined.

**Comment by shaharkazaz** [comment]
@MikeDabrowski would be great if you can update your reproduction repo, it's easier to work with it 😄

**Comment by MikeDabrowski** [comment]
@shaharkazaz 
I created the PR, could you take a look and hint me if something should be done differently ?

**Comment by shaharkazaz** [comment]
@MikeDabrowski ya sure, I have some exams now (college) so I didn't have much time to get to it, I'll let you know! Thanks!

**Comment by MikeDabrowski** [comment]
OK good luck. Glad Im over that myself haha :D

**Comment by shaharkazaz** [comment]
@MikeDabrowski merged and released in v2.4.0
